File: typescript/src/astro/galaxy-grid.ts
/**
 * Turning a **galactic position** into a place on the procedural naming grid.
 *
 * `./sector-name` maps a *sector index* to a name, and `./system-address` works in
 * *internal units* (32 per light-year, measured from the galaxy's corner). Neither
 * speaks the units a consumer actually holds: light-years with Sol at the origin,
 * as reported by the player journal, EDSM and Spansh. This module is the bridge —
 * it converts {@link GalacticPosition} to the {@link SectorGridPosition} those functions
 * want, so `sectorNameFromGridPosition` can be driven from real data.
 *
 * The three coordinate conventions on the public surface, in one place:
 *
 * | Convention | Unit | Origin | Where |
 * | --- | --- | --- | --- |
 * | {@link GalacticPosition} | light-years | Sol | journal, EDSM, Spansh, this module's input |
 * | {@link SectorGridPosition} | 1280 ly sector cubes (0–127) | galaxy corner | `sectorNameFromGridPosition` |
 * | internal units | 1/32 light-year | galaxy corner | `NamingRegionOrigin`, boxel maths |
 *
 * @packageDocumentation
 */

import type { GalacticPosition } from './galactic-position.js';
import { sectorNameFromGridPosition, type SectorGridPosition } from './sector-name.js';

/**
 * The galaxy's origin corner in galactic light-years — the point sector index
 * `0/0/0` starts at.
 *
 * @remarks
 * Sector and boxel positions are measured from this corner, which is why an
 * `id64`'s grid position has to be offset by it to become a position relative to
 * Sol. The same triple is published as `CODEX_REGION_MAP_X0` / `_Y0` / `_Z0` by
 * `./codex-region-lookup` (where it arrives with the region-cell data); it is
 * repeated here as a plain constant so converting a coordinate never pulls in the
 * region grid. A test asserts the two agree.
 */
export const GALAXY_ORIGIN: Readonly<GalacticPosition> = Object.freeze({
    x: -49985,
    y: -40985,
    z: -24105,
});

/** Edge length of one procedural sector cube, in light-years. */
export const SECTOR_EDGE_LY = 1280;

/**
 * The sector cube a galactic position falls in.
 *
 * @param position - Galactic position in **light-years, Sol at the origin** (a journal
 * `StarPos`, an EDSM/Spansh coordinate, a `ProceduralSystem.position`).
 * @returns The integer {@link SectorGridPosition} of the 1280 ly sector cube containing
 * that point, ready for {@link sectorNameFromGridPosition}.
 * @throws {RangeError} If the position lies outside the 128×128×128 sector grid
 * (i.e. outside the addressable galaxy). The message names the offending position.
 * @example
 * ```ts
 * import { sectorGridPositionFromGalacticPosition } from '@elite-dangerous-almanac/core/astro/galaxy-grid';
 *
 * // Synuefe EN-H d11-96 sits at (751, -179, -91) per EDSM
 * sectorGridPositionFromGalacticPosition({ x: 751, y: -179, z: -91 });
 * // -> { sectorX: 39, sectorY: 31, sectorZ: 18 }
 * ```
 */
export function sectorGridPositionFromGalacticPosition(
    position: GalacticPosition,
): SectorGridPosition {
    const sector = {
        sectorX: Math.floor((position.x - GALAXY_ORIGIN.x) / SECTOR_EDGE_LY),
        sectorY: Math.floor((position.y - GALAXY_ORIGIN.y) / SECTOR_EDGE_LY),
        sectorZ: Math.floor((position.z - GALAXY_ORIGIN.z) / SECTOR_EDGE_LY),
    };
    for (const v of [sector.sectorX, sector.sectorY, sector.sectorZ]) {
        if (!Number.isInteger(v) || v < 0 || v > 127) {
            throw new RangeError(
                `Galactic position outside the sector grid: ${JSON.stringify(position)}`,
            );
        }
    }
    return sector;
}

/**
 * The procedural sector name for a galactic position.
 *
 * The one-call form of {@link sectorGridPositionFromGalacticPosition} followed by
 * {@link sectorNameFromGridPosition} — "which sector is this point in?".
 *
 * @param position - Galactic position in **light-years, Sol at the origin**.
 * @returns The canonically-cased procedural sector name (e.g. `Synuefe`,
 * `Blae Eock`).
 * @throws {RangeError} If the position lies outside the sector grid, or the grid
 * slot has no procedurally assigned name.
 * @remarks
 * This is always the *procedural* sector. A system inside a hand-authored region
 * (Pleiades, Coalsack, …) is named after that region instead — resolve those with
 * {@link findHandAuthoredRegionAt} from `./hand-authored-regions`.
 * @example
 * ```ts
 * import { sectorNameFromGalacticPosition } from '@elite-dangerous-almanac/core/astro/galaxy-grid';
 *
 * sectorNameFromGalacticPosition({ x: 751, y: -179, z: -91 }); // -> 'Synuefe'
 * ```
 */
export function sectorNameFromGalacticPosition(position: GalacticPosition): string {
    return sectorNameFromGridPosition(sectorGridPositionFromGalacticPosition(position));
}

File: dotnet/src/EliteDangerousAlmanac/Astronomy/GalaxyGrid.cs
using System;

namespace EliteDangerousAlmanac.Astronomy;

/// <summary>The bridge from a galactic position to the naming grid.</summary>
/// <remarks>
/// <para>
/// <see cref="SectorName"/> reads a sector's place on the grid, and
/// <see cref="SystemAddress"/> works in the game's own internal units. A caller holds
/// neither. It holds light years with Sol at the origin, which is what the player journal,
/// EDSM and Spansh report. This class converts one to the other.
/// </para>
/// <para>
/// The three frames the surface uses are the galactic position in light years from Sol,
/// the sector grid in cubes of 1280 light years from the galaxy corner, and the internal
/// units of one thirty-second of a light year from the same corner.
/// </para>
/// </remarks>
public static class GalaxyGrid
{
    /// <summary>The corner of the galaxy, in light years from Sol.</summary>
    /// <remarks>
    /// Sector and boxel positions are counted from this point. The codex region map states
    /// the same corner, and a test holds the two together.
    /// </remarks>
    public static readonly GalacticPosition Origin = new(-49985, -40985, -24105);

    /// <summary>The edge of one sector cube, in light years.</summary>
    public const int SectorEdgeLy = 1280;

    /// <summary>Reads the sector cube one galactic position falls in.</summary>
    /// <param name="position">The position in light years, with Sol at the origin.</param>
    /// <returns>The sector's place on the grid.</returns>
    /// <exception cref="ArgumentNullException">The position is absent.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The position lies outside the grid of sector cubes, which is outside the galaxy the
    /// addressing reaches.
    /// </exception>
    public static SectorGridPosition ToSectorGridPosition(GalacticPosition position)
    {
        if (position is null) throw new ArgumentNullException(nameof(position));

        double x = Math.Floor((position.X - Origin.X) / SectorEdgeLy);
        double y = Math.Floor((position.Y - Origin.Y) / SectorEdgeLy);
        double z = Math.Floor((position.Z - Origin.Z) / SectorEdgeLy);
        if (!IsAxis(x) || !IsAxis(y) || !IsAxis(z))
        {
            throw new ArgumentOutOfRangeException(
                nameof(position),
                position,
                "The position lies outside the grid of sector cubes.");
        }

        return new SectorGridPosition((int)x, (int)y, (int)z);

        static bool IsAxis(double value) => value >= 0 && value <= 127;
    }

    /// <summary>Reads the procedural sector name of one galactic position.</summary>
    /// <param name="position">The position in light years, with Sol at the origin.</param>
    /// <returns>The sector name, such as <c>Synuefe</c> or <c>Blae Eock</c>.</returns>
    /// <exception cref="ArgumentNullException">The position is absent.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The position lies outside the grid, or the generator gives that place no name.
    /// </exception>
    /// <remarks>
    /// The answer is always the procedural sector. A system inside a region the game names
    /// by hand carries that region's name instead.
    /// </remarks>
    public static string ToSectorName(GalacticPosition position) =>
        SectorName.FromGridPosition(ToSectorGridPosition(position));
}
